File: src/main/java/hiiragi283/material/api/capability/item/HiiragiItemHandler.java
package hiiragi283.material.api.capability.item;

import hiiragi283.material.api.capability.IOControllable;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.NonNullList;
import net.minecraftforge.items.ItemStackHandler;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class HiiragiItemHandler extends ItemStackHandler implements IOControllable {

    @NotNull
    private final IOControllable.Type ioType;
    @Nullable
    private final TileEntity tile;

    public HiiragiItemHandler(int size) {
        this(size, Type.GENERAL);
    }

    public HiiragiItemHandler(int size, @NotNull IOControllable.Type ioType) {
        this(size, ioType, null);
    }

    public HiiragiItemHandler(int size, @NotNull IOControllable.Type ioType, @Nullable TileEntity tile) {
        super(size);
        this.ioType = ioType;
        this.tile = tile;
    }

    @Override
    protected void onContentsChanged(int slot) {
        if (tile != null) {
            tile.markDirty();
        }
    }

    //    Custom    //

    public void clear() {
        this.stacks = NonNullList.withSize(this.getSlots(), ItemStack.EMPTY);
    }

    public boolean isEmpty() {
        return this.stacks.stream().allMatch(ItemStack::isEmpty);
    }

    //    IOControllable    //

    @Override
    public @NotNull Type getIOType() {
        return ioType;
    }
}

File: src/main/kotlin/hiiragi283/material/api/capability/item/HiiragiItemHandler.kt
package hiiragi283.material.api.capability.item

import hiiragi283.material.api.capability.IOControllable
import hiiragi283.material.util.HiiragiNBTUtil
import hiiragi283.material.util.getStringOrNull
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraftforge.items.ItemStackHandler

open class HiiragiItemHandler(
    size: Int = 1,
    override var ioType: IOControllable.Type = IOControllable.Type.GENERAL,
    val tile: TileEntity? = null
) : ItemStackHandler(size), IOControllable {

    override fun onContentsChanged(slot: Int) {
        tile?.markDirty();
    }

    //    INBTSerializable    //

    override fun serializeNBT(): NBTTagCompound =
        super.serializeNBT().also { tag: NBTTagCompound -> tag.setString(HiiragiNBTUtil.IO_TYPE, ioType.name) }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        super.deserializeNBT(nbt)
        nbt.getStringOrNull(HiiragiNBTUtil.IO_TYPE)?.let { name: String -> ioType = IOControllable.Type.valueOf(name) }
    }

}